[1주차 자동차 경주 게임] 리뷰 부탁드립니다.

java-saeng opened the pull request:
안녕하십니까 ! 
이와 같은 미션을 통해 패키지 구조부터 하나하나 코드를 처음 작성해보았습니다.
부족한 점도 많고, 잘못된 점도 많겠지만 최선을 다해 작성해보았습니다.

아래의 글들은 저를 힘들게 했던 문제들이었고, 해결하기 위해 어떻게 접근하였는지, 해결했다면 어떻게 해결했는지 적어보았습니다. 또한, 왜 이렇게 코드를 작성했는지도 작성했습니다.

읽어주셔서 감사합니다 !

# 문제 발생 및 해결

## 무한 반복(InputView.class)

![image](https://user-images.githubusercontent.com/62413589/156744422-e9797b61-fc18-4759-82ae-1497f3de4e6a.png)



무한 반복으로 인해 `StackOverFlowError` 가 발생함

### 이렇게 짠 이유

carNum을 try 블럭 밖에서 받으면 `InputMismatchException` 을 catch 하지 못함

- try 블럭 안에서 수를 입력받음
- `StackOverFlowError` 발생
- 디버깅을 해보면서 이유를 찾는 중에 nextInt는 개행 문자를 제거하지 않는 것이 기억남
- nextLong이나 nextInt나 개행을 포함하지 않기 때문에 
sc.nextLong을 통해 개행이 계속 입력됨
- 계속해서 지나치게 되고, carNum은 0L로 초기화되어 계속 catch됨

### 해결 방법

sc.nextLine()을 통해 입력받은 뒤, parse를 이용하여 숫자로 변환

## input에 대해 Exception Unit test(InputView.class)

```java
public long inputTotalCarNum(){
        System.out.println("자동차 대수는 몇 대 인가요?");
        long carNum = 0L;
        try {
            carNum = Long.parseLong(sc.nextLine());
            if (carNum <= 0)
                throw new IllegalArgumentException();
        } catch (IllegalArgumentException | InputMismatchException exception) {
            System.out.println("자연수를 입력해주세요");
            inputTotalCarNum();
        }
        return carNum;
    }

    public int inputTotalTrial(){
        System.out.println("시도할 횟수는 몇 회 인가요?");
        int trial = 0;
        try {
            trial = Integer.parseInt(sc.nextLine());
            if (trial <= 0)
                throw new IllegalArgumentException();
        } catch (IllegalArgumentException | InputMismatchException exception) {
            System.out.println("자연수를 입력해주세요");
            inputTotalTrial();
        }
        return trial;
    }
```

입력받을 시, 문자열이나 0 또는 음수에 관해 예외처리를 하고 싶었음

### 왜 Custom Exception을 사용하지 않았나요?

- “중복된” 예외처리가 많을 경우 Custom Exception이 좋다고 생각함
- 가독성은 Custom Exception이 좋음
- 하지만 해당 미션의 예외처리는 문자열과 음수 판단으로 충분히 Java의 내장 Exception을 통해 나타낼 수 있다고 생각
- 또한, Custom Exception을 사용할 시 비용이 발생할텐데,
비용이 드는 만큼 가독성 차이가 미묘하다고 생각이 들어 Custom Exception 사용하지 않음

### 그래서 Exception test는 어떻게 하셨나요?

- 결국엔 test를 하지 못했습니다.

```java
//given
InputStream inputStream = new ByteArrayInputStream("sss".getBytes());
System.setIn(inputStream);
Scanner sc = new Scanner([System.in](http://system.in/));

//when
InputView iv = new InputView(sc);

//then
org.assertj.core.api.Assertions.assertThat(iv.inputTotalCarNum())
        .isEqualTo("자연수를 입력해주세요");
```

- System.in은 클라이언트가 입력하는 input 값을 InputStream으로 담는 역할을 함
- 따라서 직접 inputStream을 생성하여 System.in 대신에 inputStream을 넣음

![image](https://user-images.githubusercontent.com/62413589/156744584-48a232ad-3baf-4a83-bcba-877dc0f84ea9.png)


- 테스트를 실행 시 1 → 2 → 3 순으로 가게 되는데, 이 때 3에서 NoSuchElementException이 발생
- 해당 테스트를 어떻게 작성해햘지 어려움을 겪음

## 재귀로 인해 버그 발생(InputView.class)

![image](https://user-images.githubusercontent.com/62413589/156744604-875d2ce5-28e6-4f9a-a7f0-ba6519d9d3ad.png)

처음 코드이다.

### 문제

- inputTotalTrial()이 재귀로 돌아가게 됨
- 음수랑 문자열을 입력한 뒤, 자연수를 입력하게 되면
- 재귀로 인해 결국 trial이 0으로 반환됨

### 해결

![image](https://user-images.githubusercontent.com/62413589/156744649-ba152a07-0422-4db3-b3dd-b8148639ffad.png)

- inputTotalTrial() 도 리턴하면 재귀가 발생하지 않게 됨

# 코드를 짠 이유(당위성?)

## State Pattern을 왜 사용했나요?

- 미션에는 주사위의 눈에 따라 자동차가 움직이고, 움직임은 앞으로 가는 것과 정지 뿐
- 움직일 때 length를 1 늘려주면 바로 완성됨
- 하지만 추후에 자동차가 점프를 하거나, 뒤로 갈 수 있다고 생각
    - 확장 가능성
    - “움직인다” 라는 같은 행동을 취하지만, 그 행동은 “상태”에 따라 달라지게 되므로 State Pattern을 사용할 수 있다고 생각
- concrete class인 Movable, Unmovable 클래스가 State 인터페이스를 구현함으로써 설계
- 따라서 주사위의 숫자에 따라 Car의 State를 변경함
- Car class의 move메서드 하나로 자동차를 움직이거나 정지할 수 있음
    - 점프나 뒤로 갈 수 있는 기능이 생길 시 State를 구현하는 concrete class를 만들면 됨

## 패키지 구성은 왜 이렇게 했나요?
- 하나의 클래스에 하나의 책임(SRP)를 지키기 위해 아래와 같이 구성했습니다
- RacingApplication
    - 클라이언트으로부터 게임을 시작한다는 명령을 받는 책임을 가짐
- engine
    - RacingApplication이 명령을 받고 나서, 게임을 “실행”하는 책임을 가짐
- controller
    - 차를 관리하는 책임을 가짐
- function
    - 주사위를 굴리는 것은 하나의 “책임”이므로 분리시킴
- view
    - 클라이언트에게 보여지는 책임을 가짐
        - 입력 값과 출력 값을 보여줌
- model
    - Car에 대한 정보를 담은 클래스

# 결과
![image](https://user-images.githubusercontent.com/62413589/156745764-eb6a35d7-3681-433a-9cc8-349214e3b917.png)

   
# 다이어그램
![image](https://user-images.githubusercontent.com/62413589/156744807-2b7844bf-acdf-44c8-9c66-56f75837a1b0.png)

java-saeng committed:
first commit

패키지 구조 설정

java-saeng committed:
feat : Car 도메인 생성

java-saeng committed:
test : Car 도메인 test

move시 길이 변화 확인 및 식별자 확인

java-saeng committed:
feat : InputView 생성

java-saeng committed:
feat : ResultView 생성

자동차가 이동한 길이에 따라 "-" 출력

java-saeng committed:
feat : 자동차 등록 기능

입력된 갯수에 따라 자동차를 등록

java-saeng committed:
test : 자동차 등록 테스트

입력된 갯수에 따라 자동차가 잘 등록되는지 확인

java-saeng committed:
feat : 자동차 움직임 구현

getRandomNumber : 랜덤 숫자 받아오기
checkMoveOrStop : 움직일 수 있는지 판별
moveCar : 자동차를 움직임

java-saeng committed:
feat : RacingApplication 구현

게임을 시작하기 위해 객체 생성
1. 자동차 개수, 시도 횟수 입력
2. 자동차 등록
3. 각 자동차 움직이기
4. 모든 자동차의 상태 출력

java-saeng committed:
feat : 자동차 state 출력 변경

자동차 state를 출력할 때 "carNum : 이동상태"로 변경

java-saeng committed:
refactor : RacingApplication 책임 분리

RacingApplication의 책임을 RacingGame으로 변경
RacingApplication의 책임은 클라이언트로부터 게임을 시작하는 명령을 받음
RacingGame은 게임을 시작하는 책임을 가지고 있음

java-saeng committed:
chore : 변수에 private

java-saeng committed:
feat : State 패턴을 사용하여 자동차의 움직임 추상화

자동차의 움직임에 확장성을 고려하면서 Sate 패턴을 적용함
State 인터페이스를 두어 Movable, UnMovable이라는 concrete class를 둠

java-saeng committed:
test : state에 따라 움직임 검사

java-saeng committed:
feat : InputView 예외 처리

0을 포함한 음수, 문자열이 입력될 경우 예외 처리 추가

Hongdonggeon commented on src/main/java/model/State.java:
```diff
@@ -0,0 +1,5 @@
+package model;
+
+public interface State {
```
확장 가능성을 생각하여 인터페이스를 구현하였네요!! 각 다른 상태마다 객체를 만들어서 Car에서 갈아끼우는 것도 신선합니다!! State 패턴 좋은 것 배워갑니다 👍

java-saeng replied:
감사합니다 :)

hyukjin-lee replied:
state 를 적용할 다른 케이스를 생각해서 좀 더 구체적인 인터페이스명을 사용해보면 어떨까요? MoveState?

Hongdonggeon commented on src/main/java/function/RandomNumberMaker.java:
```diff
@@ -0,0 +1,10 @@
+public class RandomNumberMaker {
+
+    public static int getRandomNumber(int range) {
+        return new Random().nextInt(range);
```
getRandomNumber 가 호출될 때마다 새로운 Random객체가 만들어지지 않아도 될 것 같아요!!
그리고 Random값의boundary를 파라미터로 받은 이유가 있을까요??

java-saeng replied:
1. getRandomNumber 가 호출될 때마다 새로운 Random객체가 만들어지지 않아도 될 것 같아요!!

동건님 말씀대로 getRandomNumber를 호출할 때마다 Random 객체를 만들지 않는 것이 좋다고 생각이 듭니다
Random객체를 싱글톤 패턴을 이용하여 구현하면 리소스 부분에서 낭비를 줄일 수 있을 것 같습니다!! :)

2. Random값의 boundary를 파라미터로 받은 이유가 있을까요??

제가 생각하기에는 나중에 10에서 다른 숫자로 바꾸고 싶은 "변경 사항"을 생각해봤습니다.
차를 움직이게 하는 "CarController"에서 수정해주는 것이 좋다고 생각했습니다.

Hongdonggeon replied:
```java
private static Random random = new Random();
private static final int RANDOM_BOUNDARY = 10;

public static int getRandomNumber() {
   return random.nextInt(RANDOM_BOUNDARY);
}
```
이라고 해줄 수 있을 것 같아서요!! 
랜덤 값을 생성해주는 역할을 하는 것 같은데, 랜덤값의 범위도 랜덤값이 생성해주는 곳에서 알아야 하지 않을까 싶네요!

java-saeng replied:
"랜덤값의 범위도 랜덤값이 생성해주는 곳에서 알아야 하지 않을까 싶네요"
이 부분이 맞는 것 같습니다. 생각해보니 CarController가 랜덤값의 범위를 정한다면 결국 Random값을 생성해주는 곳에 의존하게 되어 결합도가 높아지겠네요!!
결합도를 줄이는 방법 너무 좋습니다 감사합니다!

Hongdonggeon commented on src/main/java/model/Car.java:
```diff
@@ -2,22 +2,28 @@
 
-    private String name;
-    private int position = 0;
+    long number;
```
number를 Car를 식별해주는 멤버로 사용하고 있는 것 같아요!! 접근지정자와 final 키워드를 붙여주는 것에 대해 어떻게 생각하실까요?

saint6839 replied:
접근제한자는 가능한 가장 좁은 범위로 선언해주는것이 좋다고 합니다. number, length, state 같은 경우에는 private로 선언해주면 어떨까요?

hyukjin-lee replied:
의미가 불분명한 number 보다는 식별이 가능한 값을 표현하는 id 도 괜찮을 것 같습니다 ~

Hongdonggeon commented on src/main/java/view/ResultView.java:
```diff
@@ -1,21 +1,16 @@
+    public static void printAllCarsState(long carNum, int len) {
+        System.out.print("car" + carNum + " : ");
+        for (int i = 0; i < len; i++) {
+            System.out.print("-");
```
현재 출력하는 용도로 지금 --- 이렇게 보여주고 있지만, 나중에는 다른 기호로 바뀔수도 있을 것 같아요!! 이 기호들 뿐만 아니라 InputView와 ResultView의 입출력 메세지, 매직넘버들을 상수화 해주면 좋을 것 같아요

Hongdonggeon commented on src/main/java/controller/CarController.java:
```diff
@@ -0,0 +1,39 @@
+
+import java.util.List;
+
+public class CarController {
```
전체적으로 Car들을 관리해주는 역할을 하고 있는 것 같아요! 
제가 생각하는 Controller의 역할과는 조금 다르다고 생각이 듭니다. 
제가 생각하는 Controller의 역할은 사용자가 입력한 요청값을 받아, 요청값을 가지고 비즈니스 로직을 실행하고 필요한 데이터를 가공해서 View에게 보내주는 역할이라고 생각을 합니다. 
저는 CarController보다 Cars같은 이름이 더 어울린다고 생각이 드는데 현우님은 어떻게 생각하실까요??
이 자료 참고하시면 좋을 것 같습니다!
https://tecoble.techcourse.co.kr/post/2020-05-08-First-Class-Collection/

java-saeng replied:
Car를 관리한다는 생각으로 CarController라고 했는데 Controller라는 통용된 언어가 있기에 뜻이 혼동될 수 있을 것 같습니다
앞으로 이름을 지을 때 더욱 신중히 지어야함을 느꼈습니다
제가 작성한 CarController가 동건님께서 말씀하신 일급 컬렉션에 정확하게 부합하네요!!
일급 컬렉션에 대해 알게 됐네요 감사합니다:)

Hongdonggeon commented on src/main/java/engine/RacingGame.java:
```diff
@@ -0,0 +1,47 @@
+    public RacingGame() {
+        cars = new ArrayList<>();
+        carController = new CarController();
+        iv = new InputView();
```
InputView는 인스턴스로 관리하고 ResultView는 유틸리티 클래스로 선언한 이유가 궁금하네요!! 저도 View를 구현을 할 때, 고민이 많이 되는데요! View 같은 경우에 상태가 없어서 유틸리티 클래스로 구현을 해도 될 것 같지만, 유틸성 클래스와는 색깔이 조금 다른 느낌도 있어서요!! 정답은 없겠지만, 저는 이번에 View 구현을 할 때 싱글턴 패턴을 사용해보았어요!! 현우님이 InputView와 OutputView를 다르게 구현한점이 궁금하네요!

saint6839 replied:
InputView 클래스에서 throw된 예외가 catch되는 부분이 보이지 않는 것 같아요! 리팩토링하시면서 예외 처리를 View에서 할지 Controller에서 할 지 한번 고민해보시는것도 좋은 경험이 될 것 같아요!

java-saeng replied:
RacingGame 클래스는 RacingApplication으로부터 게임 명령 "실행"을 받는 클래스로 생각했습니다. 그래서 Scanner라는 상태를 RacingGame에 두지 않고 InputView에 두었습니다. 따라서 상태가 존재하는 InputView는 유틸성 클래스로 구현하지 않았습니다!!
동건님 말씀대로 RacingGame에서 Scanner를 InputView에 인자로 전달하여 InputView와 ResultView 모두 유틸성 클래스로 관리하는 방법도 좋을 것 같습니다

Hongdonggeon commented on src/main/java/controller/CarController.java:
```diff
@@ -0,0 +1,39 @@
+    }
+
+    private State checkMoveOrStop(Car car) {
+        if (getRandomNumber() >= MOVE_CONDITION) {
```
랜덤값 때문에 checkMoveOrStop 메서드를 테스트 코드 작성하기 어려웠을 거예요!!
https://tecoble.techcourse.co.kr/post/2020-05-07-appropriate_method_for_test_by_parameter/
https://tecoble.techcourse.co.kr/post/2020-05-17-appropriate_method_for_test_by_interface/
자료 참고하시면 좋을 것 같아요!! 하지만 인터페이스를 분리하는게 무조건 좋은 방법은 아니라고 생각이 듭니다!! 의심하면서 보면 좋을 것 같아요!! 현우님은 생각이 다르실 수 도 있구요!

java-saeng replied:
strategy 패턴을 이용해서 테스트 코드를 작성할 수 있군요,, 테스트 커버리지가 높아지는 대신에 generate메서드에 있는 리턴 값을 수정해주어야 하네요
트레이드 오프를 생각해서 알맞은 설계가 중요한 것 같습니다.
인터페이스를 통해서 랜덤값 테스트를 이런 식으로 할 수 있다는 새로운 사실 을 알게 해주셔서 감사합니다!!

saint6839 commented on src/main/java/controller/RacingApplication.java:
```diff
@@ -1,24 +0,0 @@
-import java.util.List;
-import java.util.Scanner;
-
-public class RacingApplication {
```
Application 클래스에서는 그저 실행해달라는 요청만 보내고, 요청에 대한 처리는 다른 클래스에서 해주었으면 더 좋았겠다라는 생각이 듭니다. 현재는 입력을 받고 객체를 생성하는 등 너무 다양한 책임이 Applicaiton 클래스에 들어와 있는 것 같다고 생각합니다.

saint6839 commented on src/main/java/view/InputView.java:
```diff
@@ -1,16 +1,43 @@
+import java.util.InputMismatchException;
 import java.util.Scanner;
 
 public class InputView{
```
음수가 입력됐을 경우에 예외가 던져지도록 처리를 해주셨네요! 음수일 경우 말고도 숫자가 아닌 값이 입력 됐을 경우도 예외가 던져지도록 명시적으로 처리해주면 더 좋을 것 같습니다!

java-saeng replied:
상엽님 말씀대로 문자열을 입력했을 경우에도 어떤 부분이 잘못됐는지 알려주는 게 좋을 것 같아요 
클라이언트 입장에서는 문자열을 입력했을 때 어떤 점이 잘못됐는지 모르니까요!!

saint6839 commented on src/main/java/view/InputView.java:
```diff
@@ -1,16 +1,43 @@
+        long carNum = 0L;
+        try {
+            carNum = Long.parseLong(sc.nextLine());
+            if (carNum <= 0)
```
저도 동건님께 배운건데 if문 바디가 한 문장이더라도 중괄호 {}를 사용해주는것이 관례라고 하네요!

hyukjin-lee commented on src/main/java/controller/CarController.java:
```diff
@@ -0,0 +1,39 @@
+
+public class CarController {
+
+    private static final int NUMBER_RANGE = 10;
```
RANDOM_NUMBER_RANGE 는 어떨까요?

hyukjin-lee commented on src/main/java/controller/RacingGame.java:
```diff
@@ -1,66 +0,0 @@
-    public static final int CRITICAL_POINT = 5;
-    public static final int MAX_NAME_LENGTH = 5;
-
-    public void oneTrial(List<Car> cars){
```
메서드명은 행위를 표현하는 동사로 시작하는게 어떨까요?

hyukjin-lee commented on src/main/java/controller/RacingGame.java:
```diff
@@ -1,66 +0,0 @@
-
-    public void oneTrial(List<Car> cars){
-        for (Car car : cars){
-            check(car);
```
check 는 무엇을 체크한다는건지 알기 어려운 것 같아요~ 내부에서만 활용되는 클래스면 private 키워드를 붙이고 메서드명을 좀 더 풀어써보면 어떨까요?

hyukjin-lee commented on src/main/java/model/Movable.java:
```diff
@@ -0,0 +1,8 @@
+package model;
+
+public class Movable implements State {
```
~able 은 interface 의 접미사로 활용되는데
구현체의 네이밍인 경우에는 이 타입이 무엇이 가능한지보다는 무엇을 하는지 구체적인 단어를 활용하면 어떨까요?
예를 들어 Forward 나 Stop 처럼요

hyukjin-lee commented on src/main/java/model/Car.java:
```diff
@@ -2,22 +2,28 @@
-    private String name;
-    private int position = 0;
+    long number;
+    int length;
```
자동차의 위치를 나타내는 변수 같은데 position 은 어떨까요?

hyukjin-lee commented on src/main/java/model/Car.java:
```diff
@@ -2,22 +2,28 @@
 
-    public int getPosition(){
-        return position;
+    public long getCarNum() {
```
Num 도 줄임말이니 number, no 로 통일해보면 어떨까요?
그리고 getter 는 get{field명} 이라는 java bean 컨벤션이 있으니 따르면 좋을 것 같습니다

hyukjin-lee commented on src/main/java/model/Car.java:
```diff
@@ -2,22 +2,28 @@
 
-    public String getName() {
-        return name;
+    public int moveCount() {
```
count 를 이동시킨다 라는 뜻의 메서드명인 것 같아요~ getLength 는 어떨까요?

hyukjin-lee commented on src/main/java/model/Car.java:
```diff
@@ -2,22 +2,28 @@
 
-    public void move(){
-        position++;
+    public void changeState(State cmd) {
```
cmd 의 의미가 무엇일까요? command? 스터디 규칙은 축약형 단어를 사용하지 말자~ 입니다

sunghyuki replied:
setter 메서드를 사용하기 보다는 적절한 목적과 의도를 나타내는 메소드를 사용하신 부분이 좋았습니다~ 👍🏻

Entity 클래스에서는 절대 Setter 메소드를 만들지 않는다. 대신, 해당 필드의 값 변경이 필요하면 명확히 그 목적과 의도를 나타낼 수 있는 메소드를 추가해야 한다.(인용)

java-saeng replied:
맞습니다 setter를 무분별하게 사용하게 되면 어느 곳에서 값이 변경되는지 알기 힘드니까요!!

알아주시니 감사합니다 :)

hyukjin-lee replied:
"setter를 무분별하게 사용하게 되면 어느 곳에서 값이 변경되는지 알기 힘드니까요"
이러한 이유라면 이름이 set~ 인 메서드로 값을 변경하는것과 change~ 로 값을 변경하는 것의 차이는 무엇일까요? 
둘다 어디서든 값이 변경되는건 마찬가지인데 이름만 set이냐 change냐의 차이가 아닌가 싶어서요~

java-saeng replied:
말씀대로 set~이나 change 모두 변경하는 것은 똑같은데 그 이유에 대해서 말을 잘 못한 것 같습니다. 

제가 알기로는 setter메서드 사용을 지양하는 이유는
무분별한 setter를 사용한다면 변경할 필요가 없는 인스턴스에 대해서도  setter로 접근이 가능해져 객체의 일관성이나 안정성 보장이 힘들다.
또한, 무조건 사용하지 말라라는 것이 아닌 꼭 변경할 필요가 있다면 목적과 의도를 나타낼 수 있게 네이밍하여 사용하도록 하자. 인 것 같습니다!!

hyukjin-lee commented on src/main/java/model/Car.java:
```diff
@@ -2,22 +2,28 @@
+    }
+
+    public void move() {
+        length = state.execute(length);
```
state 의 기본값이 세팅되어있지 않은데, 이 클래스가 state 패턴을 적용한 것인지 모르는 사람은 car.move() 를 사용했을 경우에 NPE 가 발생하지 않을까요~?

hyukjin-lee commented on src/main/java/controller/CarController.java:
```diff
@@ -0,0 +1,39 @@
+        }
+    }
+
+    private State checkMoveOrStop(Car car) {
```
상태 패턴을 활용하신게 신선하네요~ 

다만 전진과 정지 뿐이고 상태의 전이가 없을 뿐더러 조건문이 복잡하지 않은 상태에서 적용하기엔 과하지 않나라는 고민도 해보시면 좋을 것 같아요.

move(int condition) 메서드에서 분기쳐서 전진, 정지를 수행하는 것과,

움직임 관련 상태를 정의하고, CarController 에서 분기쳐서 condition 체크하고 움직임 관련 상태를 바꿔주도록 구현한 다음, Car.move 에서 또 분기를 통해 상태값을 체크해서 행위를 결정하는 것(을 개선한 것이 상태 패턴 적용버전인데)

무엇이 더 복잡하고 심플한지 고민해보셔도 좋을 것 같아요. 간단하게 행위로 풀 것을 상태로 만들어서 패턴의 적용이 코드의 목적이 된건 아닌지도 고민해보시면 좋을 것 같습니다 ~ 

학습 목적으로는 좋은 시도인 것 같아요~

java-saeng replied:
처음에는 분기 처리하여 움직임을 바꿔줬습니다. 
하지만 예제 코드로만 state 패턴을 학습해서, 이번 기회에 애플리케이션에 사용해보고 싶었습니다.

말씀대로 과하다고 생각합니다. 처음이라서 더 잘해봐야지라는 생각으로 그랬던 것 같아요,, !
또한, 제가 어느정도 공부를 했는지 모르시니까 여기까지 공부해봤어요라고 표현하고 싶었습니다.

hyukjin-lee replied:
넵 의도는 충분히 이해하고 있습니다 ~

다만 저는 멘티분들이 학문을 탐구하는 것에 그치지 않고 주니어로서 일을 잘하는 좋은 개발자가 되기 위한 고민 포인트와 조언을 노파심에 계속 던져드릴거라서, 새로운 지식에 대한 학습이나 시도에 대해서는  지금처럼 당당하게 주눅 들지 마시고 그냥 코멘트에 대해 한번 생각해보고 넘어간다는 차원으로 받아들여주시면 좋을 것 같아요.

덕분에 상태 패턴도 한번 돌이켜보게 되고 저도 공부가 되어서 좋네요 ㅎㅎ

java-saeng replied:
넵!!  새로운 도전 당당하게 하겠습니다 ! 
항상 좋은 말씀 해주셔서 감사드립니다 !!

sunghyuki replied:
저도 새로운 방법을 배워가네요~~

hyukjin-lee commented on src/main/java/view/InputView.java:
```diff
@@ -1,16 +1,43 @@
+        sc = new Scanner(System.in);
+    }
+
+    public long inputTotalCarNum(){
```
코드의 포맷팅도 컨벤션의 일종입니다.
intellij 의 코드 포맷팅 단축키를 활용해서 정렬해보시면 어떨까요?

hyukjin-lee commented on src/test/java/model/CarTest.java:
```diff
@@ -1,39 +1,53 @@
-        hyundai = new Car("hyundai");
+    @Test
+    @DisplayName("차 번호가 식별자이다")
+    public void identification() throws Exception{
```
junit5 부터는 접근제어자를 생략가능해서 좀 더 가독성 있고 간결한 테스트 코드를 작성할 수 있습니다~

hyukjin-lee commented on src/test/java/model/CarTest.java:
```diff
@@ -1,39 +1,53 @@
-        kia = new Car("kia");
-        hyundai = new Car("hyundai");
+    @Test
+    @DisplayName("차 번호가 식별자이다")
```
이거는 getter 를 테스트하는 격이라서 테스트가 필요가 없을 것 같다는 생각이 듭니다.
~할 때 ~ 한다. 라고 테스트명을 작성해주시면 나중에 클래스의 기능 명세로도 활용되어 좋은 테스트가 될 것 같아요

hyukjin-lee reviewed (commented):
고생 많으셨습니다.
뒤늦게 따라오시느라 많이 힘드실텐데
너무 열정적으로 잘 따라와주셔서 감사합니다
새로운 시도도 보기 좋고 고민도 많이 하신 것 같아서 좋네요.
제 개인적으로 좋은 코드는 작성자의 고민이 담긴 코드라고 생각합니다.
고생 많으셨습니다.

java-saeng committed:
refactor : RandomNumberMaker와 CarController 의존관계

의미를 정확하게 나타내기 위해 상수명 변경
NUMBER_RANGE -> RANDOM_NUMBER_RANGE

CarController에서 RandomNumberMaker로 랜덤 값의 범위를 전달함으로써 둘의 의존관계가 생김
RandomNumberMaker는 랜덤 값을 생성하는 책임을 가지고 있기 때문에 랜덤 값의 범위는 RandomNumberMaker가 알고 있어야한다고 생각
RandomNumberMaker가 랜덤 값을 알고 있게 되면서 CarController와의 의존 관계가 없어짐

java-saeng committed:
refactor : 매직 넘버 및 리터럴 수정, 예외 명시적으로 전달

매직 넘버와 리터럴을 static final로 상수화함으로써 명시적으로 무엇을 하는지 나타냄
문자열을 입력했을 경우에도 예외 메시지를 명시함

sunghyuki commented on src/main/java/engine/RacingGame.java:
```diff
@@ -0,0 +1,47 @@
+
+    private List<Car> cars;
+    private CarController carController;
+    private InputView iv;
```
iv는 inputView의 축약어 같습니다. 의미가 불분명한 iv 보다는 의미가 분명한 inputView도 좋을 것 같습니다~

sunghyuki commented on src/main/java/engine/RacingGame.java:
```diff
@@ -0,0 +1,47 @@
+        long carNum = iv.inputTotalCarNum();
+        int trial = iv.inputTotalTrial();
+
+        regist(carNum);
```
regist란 단어는 존재하지 않는 것 같아 보입니다. register 라는 동사는 사용하는 건 어떨까요?

java-saeng replied:
그렇네요,, 어디서 이런 단어를 들었을까요 
감사합니다!!

java-saeng committed:
chore : 메서드, 변수 명 변경

iv 같은 축약어 사용 X inputView 변경
regist라는 단어 존재 X register 변경
